plugins/parsers/bindata/parser.go

```diff
@@ -0,0 +1,216 @@
+				if (timeFormat == "unix_ms" || timeFormat == "unix_us" || timeFormat == "unix_ns") &&
+					fieldType != reflect.TypeOf((*int64)(nil)).Elem() {
+					return nil, fmt.Errorf(`invalid time type, must be int64`)
+				}
+			}
+
+			// Overwrite non-string and non-padding field size
+			if fieldType.Name() != "string" {
+				fields[i].Size = uint(fieldType.Size())
+			}
+		}
+	}
+
+	return &BinData{
+		metricName:     metricName,
+		timeFormat:     timeFormat,
+		endiannes:      endiannes,
+		byteOrder:      byteOrder,
+		stringEncoding: stringEncoding,
+		fields:         fields,
+		DefaultTags:    defaultTags,
+	}, nil
+}
+
+// SetDefaultTags implements Parser.SetDefaultTags()
+func (binData *BinData) SetDefaultTags(tags map[string]string) {
+	binData.DefaultTags = tags
+}
+
+// Parse implements Parser.Parse()
+func (binData *BinData) Parse(data []byte) ([]telegraf.Metric, error) {
+
+	fields := make(map[string]interface{})
+	var offset uint = 0
+
+	for _, field := range binData.fields {
+		if offset > uint(len(data)) || offset+field.Size > uint(len(data)) {
+			return nil, fmt.Errorf("invalid offset/size in field %s", field.Name)
+		}
+		if field.Type != "padding" {
+			fieldBuffer := data[offset : offset+field.Size]
+			switch field.Type {
+			case "string":
+				fields[field.Name] = string(fieldBuffer)
+			default:
+				fieldValue := reflect.New(fieldTypes[field.Type])
+				byteReader := bytes.NewReader(fieldBuffer)
+				binary.Read(byteReader, binData.byteOrder, fieldValue.Interface())
+				fields[field.Name] = fieldValue.Elem().Interface()
+			}
+		}
+		offset += field.Size
+	}
+
+	metricTime, err := binData.getTime(fields)
+	if err != nil {
+		return nil, err
+	}
+
+	metric, err := metric.New(binData.metricName, binData.DefaultTags, fields, metricTime)
+	if err != nil {
+		return nil, err
+	}
+
+	return []telegraf.Metric{metric}, err
+}
+
+// ParseLine implements Parser.ParseLine()
+func (binData *BinData) ParseLine(line string) (telegraf.Metric, error) {
+	return nil, fmt.Errorf("BinData.ParseLine() not supported")
+}
+
+func (binData *BinData) getTime(fields map[string]interface{}) (time.Time, error) {
+	nilTime := new(time.Time)
+	metricTime := time.Now()
+	timeValue := fields[timeKey]
+	if timeValue != nil {
+		var err error
+		switch binData.timeFormat {
+		case "unix":
+			if _, ok := timeValue.(int32); !ok {
+				return *nilTime, fmt.Errorf("invalid time type, must be int32")
+			}
+			metricTime, err = internal.ParseTimestamp(binData.timeFormat, int64(timeValue.(int32)), timezone)
+		case "unix_ms", "unix_us", "unix_ns":
+			if _, ok := timeValue.(int64); !ok {
+				return *nilTime, fmt.Errorf("invalid time type, must be int64")
+			}
+			metricTime, err = internal.ParseTimestamp(binData.timeFormat, int64(timeValue.(int64)), timezone)
+		default:
+			return *nilTime, fmt.Errorf("invalid time format %s", binData.timeFormat)
+		}
+		if err != nil {
+			return *nilTime, err
+		}
+		delete(fields, timeKey)
+	}
+	return metricTime, nil
```

Comment:
This should be simplified a bit
```suggestion
	t, found := fields[timeKey]
	if !found {
		return time.Now()
	}
	delete(fields, timeKey)
	
	switch binData.timeFormat {
	case "unix":
		tval, ok := t.(int32)
		if !ok {
			return nil, fmt.Errorf("invalid time type %T, must be int32", t)
		}
		return internal.ParseTimestamp(binData.timeFormat, int64(tvalue), timezone)
	case "unix_ms", "unix_us", "unix_ns":
		tval, ok := t.(int64)
		if !ok {
			return nil, fmt.Errorf("invalid time type %T, must be int64", t)
		}
		return internal.ParseTimestamp(binData.timeFormat, int64(tval), timezone)
	}
	return nil, fmt.Errorf("invalid time format %q", binData.timeFormat)
```